src/main/scala/com/microsoft/hyperspace/actions/RefreshDeleteAction.scala

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import scala.collection.immutable.HashSet
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.functions._
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEvent, RefreshDeleteActionEvent}
+
+/**
+ * Refresh index by removing index entries from any deleted source data file.
+ * Note this Refresh Action only fixes an index w.r.t deleted source data files
+ * and does not consider new source data files (if any).
+ * If some original source data file(s) are removed between previous version of index and now,
+ * this Refresh Action updates the index as follows:
+ * 1. Deleted source data files are identified;
+ * 2. Index records' lineage is leveraged to remove any index entry coming
+ *    from those deleted source data files.
+ *
+ * @param spark SparkSession
+ * @param logManager Index LogManager for index being refreshed
+ * @param dataManager Index DataManager for index being refreshed
+ */
+class RefreshDeleteAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager)
+    extends RefreshActionBase(spark, logManager, dataManager)
+    with Logging {
+
+  final override protected def event(appInfo: AppInfo, message: String): HyperspaceEvent = {
+    RefreshDeleteActionEvent(appInfo, logEntry.asInstanceOf[IndexLogEntry], message)
+  }
+
+  /**
+   * Validate index has lineage column and it is in active state for refreshing and
+   * there are some deleted source data file(s).
+   */
+  final override def validate(): Unit = {
+    super.validate()
+    if (!previousIndexLogEntry.hasLineageColumn(spark)) {
+      throw HyperspaceException(
+        s"Index refresh (to handle deleted source data) is" +
+          " only supported on an index with lineage.")
+    }
+
+    if (deletedFiles.isEmpty) {
+      throw HyperspaceException("Refresh aborted as no deleted source data file found.")
+    }
+  }
+
+  /**
+   * For an index with lineage, find all the source data files which have been deleted,
+   * and use index records' lineage to mark and remove index entries which belong to
+   * deleted source data files as those entries are no longer valid.
+   */
+  final override def op(): Unit = {
+    logInfo(
+      "Refresh index is updating index by removing index entries" +
+        s" corresponding to ${deletedFiles.length} deleted source data files.")
```

Comment:
The space is needed as this is a single sentence, without space it is gonna be ".. entriescorresponding ..". I am gonna move the space from beginning of second string to the end of first string, but it can be completely deleted.

Reply:
Yea, I meant to put it to the previous line.